# Тема 3. Основи Python: синтаксис, типи даних, введення і виведення

# 1. Типова структура програми Python

Програма на мові Python – це текстовий файл з розширенням `.py`, що містить набір інструкцій, які інтерпретатор виконує послідовно, рядок за рядком, згори донизу. Хоча прості програми можуть складатися лише з кількох рядків коду, для кращої організації та читабельності варто дотримуватися певної структури.

### Загальні принципи організації коду

**Ключові елементи структури:**

* **Коментарі:** Важлива частина будь-якої програми. Це текстові пояснення, які ігноруються інтерпретатором, але слугують для пояснення логіки коду програмістам. У Python однорядкові коментарі починаються із символу `#`.

In [1]:
# Це коментар - текст після # ігнорується інтерпретатором
# Коментарі використовуються для пояснення коду
# x = 1
# print(x)

In [2]:
"""
Це багаторядковий коментар
Він може займати декілька рядків
"""
'''
y = 2
print(x + y)
'''
print('Привіт, Світ!')

Привіт, Світ!


**Інструкції (Statements):** Кожен рядок коду, що виконує певну дію (наприклад, присвоєння значення змінній, виклик функції), є інструкцією.

Типова структура простої програми може виглядати так:
1.  **Імпорт модулів (за потреби):** Підключення зовнішніх бібліотек, які розширюють функціональність.
2.  **Оголошення констант та глобальних змінних (за потреби):** Визначення даних, що будуть доступні в усій програмі.
3.  **Визначення функцій (за потреби):** Створення блоків коду для повторного використання.
4.  **Основний блок виконання:** Головна логіка програми, що виконується при запуску файлу.

Для найпростіших скриптів програма може складатись лише з основного блоку виконання.
Python не потребує обов’язкової функції `main`, як у деяких інших мовах. Програма починає виконання з першої інструкції у файлі.
Отже, програма Python може виконуватися як скрипт (без функції `main`) або з використанням умовної конструкції `if __name__ == "__main__":` для розмежування основного коду від імпортованого.

In [3]:
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Документація модуля (docstring)
Опис призначення програми
"""

# Імпорт необхідних модулів
import math

# Константи (пишуться великими літерами)
PI = math.pi
RADIUS = 10

# Визначення функцій
def calculate_area(radius):
    """Обчислює площу кола"""
    return PI * radius ** 2

# Основний код програми
def main():
    """Головна функція програми"""
    area = calculate_area(RADIUS)
    print(f"Площа кола: {area:.2f}")

# Точка входу в програму
if __name__ == "__main__":
    main()

Площа кола: 314.16


Рядки
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
```
не впливають на логіку програми, але забезпечують правильне виконання та інтерпретацію файлу в різних середовищах.
1. **Shebang-рядок**: `#!/usr/bin/env python3`
Призначення: Цей рядок називається "shebang" (від "sharp" # і "bang" !) і використовується в Unix-подібних системах (Linux, macOS) для вказівки, яким інтерпретатором виконувати файл.
2. **Кодування**: `# -*- coding: utf-8 -*-`
Призначення: Цей рядок вказує Python, в якому кодуванні записаний файл з кодом.

**Історичний контекст**:
- У Python 2 за замовчуванням використовувалось ASCII-кодування
- Якщо в коді були символи поза ASCII (кирилиця, емодзі, спецсимволи), виникали помилки
- Цей рядок дозволяв явно вказати UTF-8 кодування

**Сучасний стан**:

- У Python 3 UTF-8 є кодуванням за замовчуванням
- Цей рядок більше не є обов'язковим для більшості випадків
 -Залишається корисним для явного вказання кодування або сумісності з Python 2

# 2. Синтаксичні особливості Python

**Блоки коду та відступи (Indentation):** На відміну від багатьох інших мов, де для визначення блоків коду використовуються фігурні дужки `{}`, у Python для цього слугують **відступи** (пробіли або табуляція). Це одна з ключових синтаксичних особливостей мови. Всі інструкції в межах одного блоку (наприклад, тіло циклу або функції) повинні мати однаковий рівень відступу. Стандартною практикою є використання 4 пробілів для одного рівня вкладеності. Ця особливість робить код більш читабельним і примушує програмістів дотримуватися єдиного стилю оформлення.

In [4]:
# Приклад використання відступів
number = 15
if number > 10:
    print("Число більше 10")  # Відступ 4 пробіли
    print('Це другий рядок блоку')  # Той самий відступ
print("Цей блок не входить до блоку умови")

Число більше 10
Це другий рядок блоку
Цей блок не входить до блоку умови


| Особливість | Опис | Приклад |
|-------------|------|---------|
| **Відступи** | Визначають блоки коду (4 пробіли) | `    print("Відступ")` |
| **Двокрапка** | Завершує заголовки блоків | `if x > 0:` |
| **Регістрочутливість** | `Name` ≠ `name` | `variable` ≠ `Variable` |
| **Без крапки з комою** | В кінці рядка не потрібна | `x = 10` |

Інструкції в Python зазвичай розміщуються по одній на рядок. Для продовження довгого рядка на наступний використовується зворотна коса риска `\` або круглі дужки.
Неявне продовження рядків відбувається автоматично всередині круглих `()`, квадратних `[]` або фігурних `{}` дужок. Python розуміє, що вираз ще не завершено, поки не буде закрита відповідна дужка, тому дозволяє переносити код на наступні рядки без додаткових символів. Цей спосіб є більш читабельним та рекомендованим порівняно з використанням зворотної косої риски.

Тепер щодо прикладів, як це працює на практиці:
1. Використання круглих дужок для довгих виразів:

In [5]:
first_variable = 1
second_variable = 2
third_variable = 3
fourth_variable = 4
fifth_variable = 5

# Довгий математичний вираз
result = (first_variable + second_variable +
          third_variable + fourth_variable +
          fifth_variable)

# Довгий виклик функції
print("Це дуже довгий рядок",
      "який ми хочемо розбити",
      "на кілька рядків для кращої читабельності")

Це дуже довгий рядок який ми хочемо розбити на кілька рядків для кращої читабельності


2. Використання квадратних дужок для списків:

In [6]:
# Довгий словник
person = {
    "name": "John Doe",
    "age": 30,
    "city": "Kyiv",
    "occupation": "Developer"
}

4. Порівняння з використанням зворотної косої риски:

In [7]:
# Менш читабельний спосіб
result = first_variable + second_variable + \
         third_variable + fourth_variable + \
         fifth_variable

# Більш читабельний спосіб (з дужками)
result = (first_variable + second_variable +
          third_variable + fourth_variable +
          fifth_variable)

Ключова відмінність полягає в тому, що всередині дужок Python автоматично розуміє, що вираз продовжується на наступному рядку, тому не потрібно додавати спеціальні символи. Це робить код чистішим та читабельним. Максимальна рекомендована довжина рядка - 80 символів.

# 3. Змінні
**Замість епіграфу**: *Коли дружина програміста народила двійню щоб не заморочуватись він назвав дітей `new_daughter1` і `new_daughter2`*

    Змінна (variable) — об'єкт програми, що має ім'я та значення.

Усі дані у комп'ютері зберігаються у пам'яті. Щоб отримати доступ до даних нам необхідно знати де саме у пам'яті вони знаходяться. Пам'ять комп'ютера поділена на комірки, і кожна комірка має свій номер. Отже у якості місця у пам'яті можна вказати її номер (адрес). На самому початку розвитку комп'ютерної техніки люди писали програми у машиних кодах, тобто у прямих інструкціях для процесора, і місцезнаходження даних так і визначали — вказуючи їх адрес у пам'яті. Це була дуже кропітка і складна робота, припуститись помилки було дуже легко. Згодом була написана програма Асемблер (Assembler, складальник), яка отримувала програму у текстовому вигляді і перекодувала її у машині коди для процесора. Програма складалась з інструкцій для процесора, але вже у більш зрозумілому для людини вигляді. А для звернення до певної комірки пам'яті замість адреси комірки почали використовувати мнемонічні імена. Це набагато полегшило процес розробки програм, а правила представлення машиних команд у текстовому вигляді назвали "мова Асемблера". Можна сказати що мова Асемблера — це перша мова програмування.

З розвитком мов програмування вони дедалі ставали усе вищого рівня, тобто такими, які ближче до людської мови ніж до команд для процесора. І для звернення до даних у пам'яті з'явилось таке поняття як змінна.

У різних мовах програмування принципи роботи зі змінними можуть дещо відрізнятись. Для початку припустимось ствердження що змінна — це певне ім'я, яке ми можемо пов'язати з певними даними, і для доступу до останніх ми можемо використовувати це ім'я.

### Змінні у Python
Щоб пов'язати дані з іменем використовують операцію присвоєння. В Python ця операція позначається знаком "дорівнює". Зліва від знаку "=" вказують ім'я (змінну), а праворуч — дані які треба пов'язати з цим ім'ям. Наприклад:

```python
quantity = 7
```
Цей вираз означає що до даних які дорівнюють числу 7 ми будемо звертатись за ім'ям `quantity`.
А читається цей вираз так: "Змінній `quantity` присвоєно значення `7`".

### Іменування змінних
Імена змінним у Python можна давати підпорядковуючись певним правилам:

- ім'я може містити букви, цифри, символ підкреслення (_)
- ім'я не може починатись з цифри
- ім'я не може співпадати з одним з ключових слів (словами які використовуються у мові програмування Python)
- імена чутливі до регістру
Спи#сок ключових слів можна дізнатись виконавши спеціальну команду:

In [8]:
help('keywords')


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 



Часто імена змінних включають у себе декілька слів. Наприклад, «ім'я користувача».

У різних мовах застосовують різні стилі іменування, ось основні:

`camelCase` — перше слово з маленької букви, усі наступні — з великої
`snake_case` — усі слова з маленької букви, розділяють символом підкреслення
`kebab-case` — усі слова з маленької букви, розділяють символом дефіса
`PascalCase` — усі слова з великої букви

Змінні в Python іменують в стилі `snake_case`. Усі слова — англійською мовою.

```python
user_name
```

Серед розробників є жарт: `«саме складне у програмуванні — назви змінних і інвалідація кеша»`. Придумати назву дійсно деколи складно. Як би ви назвали змінну, у якій зберігається кількість неоплачених замовлень від клієнтів, які мають заборгованість у минулому кварталі?

Але все ж намагайтесь давати змінним осмислені імена які відносяться до даних, на які вони вказують. З часом, вивчаючи чужий код, ви сформуєте правильні поняття іменування.

### Використання змінних
Ми можемо давати команди інтерпретатору присвоїти змінним певні значення. А щоб дізнатись значення змінної достатньо ввести її ім'я (в інтерактивній консолі або Jupiter ноутбуці) або розрукувати її через команду `print` (в скріпті)

Приклад:

In [9]:
usd_rate = 41.25
usd = 150
uah = usd_rate * usd

In [10]:
uah

6187.5

In [11]:
print(usd_rate)

41.25


У Python можна робити множинне присвоювання:

In [12]:
# Множинне присвоєння
x, y, z = 1, 2, 3         # різні значення
a = b = c = 8             # однакове значення

In [13]:
print(y)

2


In [14]:
print('a =', a)

a = 8


In [15]:
# # String Interpolation / f-Strings
print(f'{b = }')

b = 8


In [16]:
# “Old Style” String Formatting (% Operator)
print('c = %s' %(c))

c = 8


In [17]:
# “New Style” String Formatting (str.format)
print('a = {}; b = {}; c = {};'.format(a,b,c))

a = 8; b = 8; c = 8;


Коли ми присвоюємо змінній певне значення вперше, говорять що ми `проініціалізували` змінну, а така дія називається `ініціалізація`.

Щоб використовувати будь-яку змінну її треба спочатку ініціалізувати. Якщо ж ми спробуємо отримати значення непроініціалізованої змінної, то інтерпретатор повідомить нам що ми припустились помилки, або ж у термінах Python що "виникла вийняткова ситуація".

In [18]:
variable = 3
ans = variable * 10

Інформація про виняткову ситуацію починається рядком
```python
Traceback (most recent call last):
```

а як правило останній рядок пояснює нам що саме спричинило вийняткову ситуацію.

```python
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[18], line 2
      1 # variable = 3
----> 2 ans = variable * 10

NameError: name 'variable' is not defined
```

У даному випадку рядок
```python
NameError: name 'variable' is not defined
```
можна прочитати як:

    Помилка імені: ім'я 'variable' не визначене.

Тобто ми намагались отримати значення змінної a але не проініцілізували її перед цим.

Детальніше про вийняткові ситуації та що з ними робити коли вони виникають ми дізнаємось у подальшому.

### Обмін значеннями (Python особливість)

In [19]:
first = "перший"
second = "другий"
first, second = second, first
print(f"first = {first}, second = {second}")  # first = другий, second = перший

first = другий, second = перший


### Таблиця "Правила іменування змінних"

| Правило | Приклад правильного | Приклад неправильного |
|---------|--------------------|-----------------------|
| Починається з літери або `_` | `name`, `_private` | `2age`, `$money` |
| Містить літери, цифри, `_` | `user_age`, `score2` | `user-name`, `user age` |
| Регістрочутливі | `myVar` ≠ `myvar` | - |
| Не ключові слова | `variable` | `if`, `class`, `def` |
| `snake_case` стиль | `user_name`, `max_score` | `userName`, `MaxScore` |

### Константи
Деякі дані, такі як математичні постійні, ніколи не змінюються. Наприклад число "`пі`". Наближено воно завжди дорівнє `3.14` і не може змінюватись. Такі дані називають константами.

В Python немає спеціального механізму для визначення констант. Створення константи не відрізняється від створення змінної. Але є домовленість: константи прийнято іменувати буквами у верхньому регістрі.
```python
PI = 3.14159
```

# 4. Об’єктна модель Python

У мові Python **все є об'єктом** - числа, рядки, списки, навіть самі типи даних. Об'єкт - це область пам'яті, яка містить деяке значення та асоціаційовані з ним набори операцій.

`Об’єкт` – це абстракція, яка використовується для представлення даних у мові Python. Об’єкт є фундаментальним поняттям у мові програмування Python (і не тільки в Python). Об’єкт є синонімом до слова екземпляр, який містить деяке значення даних.

Будь-які операції (додавання, віднімання тощо) у мові Python виконуються над об’єктами.

Дані у вигляді об’єктів можуть бути представлені двома способами:

- вбудованими об’єктами, які предоставляються мовою Python;
- власноруч створеними об’єктами, з застосуванням конструкцій мови Python чи бібліотек розширень.

Використання вбудованих об’єктів дає наступні переваги:

- спрощується розробка програм. Не витрачаються зайві ресурси на розробку власних складних об’єктів. Достатньо використовувати наявні потужні вбудовані засоби мови Python;
- знижується кількість помилок на розробку власних об’єктів, оскільки програмний код вбудованих об’єктів є протестований;
- забезпечується достатньо висока швидкодія з обробки колекцій, списків, таблиць тощо. Це пояснюється тим, що програмний код вбудованих об’єктів є максимально оптимізований;
- можливість розробки власних складних об’єктів на базі вбудованих класів мови Python чи інтерфейсів мови C. Це в свою чергу полегшує розширюваність вже існуючого коду;
- вбудовані об’єкти реалізовані так, що вони є стандартною і незмінною частиною мови Python. Це дає перевагу зручності та високій ефективності їх використання у програмах. Власно розроблені об’єкти мають схильність змінюватись від одного випадку до іншого.

### Характеристики об'єкту

Кожен об'єкт має три характеристики:

1. **Ідентичність** - унікальне число для кожного об'єкту
2. **Тип** - визначає можливі операції та значення
3. **Значення** - дані, які зберігає об'єкт

In [20]:
# Визначення характеристик об'єкту
x = 42

In [21]:
# Значення
print(x)            # 42

42


In [22]:
# Тип об'єкту
print(type(x))      # <class 'int'>

<class 'int'>


`Ідентичність об’єкту` – це ціле унікальне і постійне (константне) число, яке встановлюється для даного конкретного об’єкту. Ідентичність об’єкту встановлюється тільки один раз при створенні об’єкту. Ідентичність даного об’єкту ніколи не змінюється після його створення. Ідентичність об’єкту асоціюється з адресою об’єкту в пам’яті.

Функція `id()` призначена для отримання значення ідентичності об’єкту. Функція повертає ціле число, яке для даного об’єкту є унікальним та постійним.

In [23]:
# Ідентичність - унікальний ідентифікатор в пам'яті
print(id(x))        # наприклад: 140706012416200

140706026899656


In [24]:
print(id(42))

140706026899656


In [25]:
a=5
print(id(a))

140706026898472


In [26]:
pa=id(a)
print(pa)

140706026898472


In [27]:
print(id(5))

140706026898472


In [28]:
print(id(6))

140706026898504


У вищенаведеному прикладі визначається значення ідентичності для об’єктів з іменами `a`, `pa`, числа `6`, числа `5`. Як видно з прикладу, числа також є об’єктами. Значення ідентичності для різних імен можуть співпадати.

З допомогою операторів `is` та `is not` можна порівнювати значення об’єктів на ідентичність.

In [29]:
a=3
b=5
print(a is b)

False


In [30]:
c = 5
print(c is b)

True


In [31]:
print(c is not a)

True


In [32]:
x = 256
y = 256
print(x is y)

True


In [33]:
x = 257
y = 257
print(x is y)

False


In [34]:
print(x == y)

True


Це приклад того, як у Python працює оператор is та внутрішня оптимізація для цілих чисел.

🔍 Що відбувається
- is перевіряє чи це один і той самий об’єкт у пам’яті, а не чи значення однакові.
- Для малих цілих чисел (зазвичай від -5 до 256) Python створює їх наперед і зберігає в спеціальному кеші. Це називається інтернування (interning).
- Тому, коли ми пишемо:
```python
x = 256
y = 256
print(x is y)  # True
```
- обидві змінні посилаються на той самий об’єкт у пам’яті.
- Але для чисел поза цим діапазоном (наприклад, 257) Python не гарантує кешування:
```python
x = 257
y = 257
print(x is y)  # False
```
- У цьому випадку створюються два різні об’єкти з однаковим значенням, тому is повертає False.
  
💡 Висновок:
- Якщо потрібно перевіряти значення, використовуємо `==`.
- Якщо потрібно перевірити, чи це той самий об’єкт, використовуємо `is`.


In [35]:
int_5 = 5
float_5 = 5.0
print(int_5 == float_5)

True


In [36]:
print(int_5 is float_5)

False


Цей приклад показує різницю між **`==`** (перевірка рівності значень) та **`is`** (перевірка ідентичності об’єктів у пам’яті) у Python. 

---

#### 🔍 Що відбувається крок за кроком

```python
int_5 = 5
float_5 = 5.0
print(int_5 == float_5)  # True
```
- **`==`** перевіряє, чи значення однакові.  
- `5` (тип `int`) і `5.0` (тип `float`) мають однакове числове значення, тому результат **`True`**.  
- Python автоматично виконує порівняння з приведенням типів, якщо це можливо.

```python
print(int_5 is float_5)  # False
```
- **`is`** перевіряє, чи це **той самий об’єкт у пам’яті**.  
- `int_5` — це ціле число (`int`), а `float_5` — число з плаваючою комою (`float`).  
- Це **різні типи** і різні об’єкти, навіть якщо значення збігаються.  
- Тому результат **`False`**.

---

#### 📌 Підсумок
- **`==`** → порівнює **значення** (може бути True навіть для різних типів, якщо значення збігаються).  
- **`is`** → перевіряє **ідентичність об’єкта** (True тільки якщо змінні вказують на один і той самий об’єкт у пам’яті).  

---


# 5. Дані та їх типи
У реальному житті ми виконуємо різноманітні дії над оточуючими нас предметами (об'єктами). Ми змінюємо їх властивості, наділяємо новими функціями. По аналогії з цим комп'ютерні програми також маніпулюють об'єктами, тільки віртуальними, цифровими. Поки що будемо називати такі об'екти даними.

Вочевидь, дані бувають різними: числа, текстова інформація, аудіозапис. Числа у свою чергу також бувають різними: цілими, дійсними, можуть мати дуже велике значення або дуже довгу дробову частину.

    Тип даних — характеристика, яка визначає множину припустимих значень, формат їхнього збереження та набір операцій (деколи також розмір виділеної пам'яті), які можна робити над даними.

### Основні типи даних
Python має багато вбудованих типів даних. Ось деякі найважливіші:

* Числа можуть бути цілими (1 і 2), з десятковими дробами (1.1 і 1.2), звичайними дробами (½ and ⅔), чи навіть комплексними.
* Логічні (Булеві) змінні приймають значеня True або False.
* Рядки (послідовності символів, наприклад 'To be or not to be' або ж цілий HTML документ)
* Байти та масиви байтів (наприклад зображення в форматі JPEG)
* Списки (впорядковані послідовності значень).
* Кортежі (впорядковані незмінні послідовності значень).
* Множини (набори унікальних значень).
* Словники (набори пар ключ-значення).

<img src="data_types.svg" alt="Типи даних" width="400"/>

  
В Python тип даних не треба вказувати явно. Покладаючись на початково присвоєне змінній значення, Python визначає якому типу воно належить і запам’ятовує його самостійно. Таку типізацію називають динамічною або неявною.

# 6. Динамічна типізація та анотації типів

### Динамічна типізація
Python є мовою з **динамічною типізацією** - тип змінної визначається автоматично при присвоєнні значення:

In [37]:
# Не потрібно оголошувати тип
x = 10          # x автоматично стає int
print(type(x))

<class 'int'>


In [38]:
x = "Python"    # тепер x стає str
print(type(x))

<class 'str'>


In [39]:
x = 3.14        # тепер x стає float
print(type(x))

<class 'float'>


Не зважаючи на це, буває корисно явно вказувати тип змінної використовуючи `анотацію типів`.
 
---

#### 📌 Що таке анотація типів
Анотація типів (type hints) — це спосіб **підказати** іншим програмістам (і інструментам, як-от IDE чи лінтери), які типи даних очікуються у змінних, аргументах функцій та значеннях, що повертаються.  
Вона **не змінює поведінку програми** під час виконання, але допомагає:
- робити код зрозумілішим;
- знаходити помилки ще до запуску;
- полегшувати автодоповнення в редакторах.

---

#### 🔹 Приклади з простими типами

```python
age: int = 25        # ціле число
price: float = 19.99 # число з плаваючою комою
product: str = "Apple" # рядок
```
Тут ми вказали, що:
- `age` має бути цілим числом (`int`);
- `price` — десятковим (`float`);
- `product` — рядком (`str`).

#### 💡 Корисно знати
- Python **не примушує** дотримуватися анотацій — це лише підказка.
- Для перевірки типів можна використовувати інструменти, як-от **mypy**:
  ```bash
  mypy script.py
  ```
---

#### Базові типи
| Тип | Приклад значення | Анотація |
|-----|------------------|----------|
| `int` | `42` | `age: int` |
| `float` | `3.14` | `pi: float` |
| `str` | `"Hello"` | `name: str` |
| `bool` | `True` | `is_active: bool` |
| `None` | `None` | `value: None` |

---

### Операції у програмуванні
`Операція` – це виконання певних дій над даними. Дані, над якими виконують дій, називають `операндами`. Саму дію виконує `оператор` – спеціальний інструмент.

Наприклад додавання чисел:
```python
2 + 3
```
Тут операцією виступає додавання, операндами — числа `2` і `3`, оператором — знак "`+`".

Зауважимо, що `один і той же оператор може виконувати різні дії (операції) залежно від типу даних операндів`.

In [40]:
number1 = 20
number2 = 30
res_number = number1 + number2
print(res_number)

50


In [41]:
string1 = '20'
string2 = '30'
res_string = string1 + string2
print(res_string)

2030


# 6. Числа
Python підтримує як цілі, так і дробові числа. Вказувати тип числа явно не потрібно, Python сам визначить його за наявністю чи відсутністю десяткової крапки.

# 6.1 Цілі числа (int)
Числа в Python нічим не відрізняються від звичайних чисел. 

In [42]:
# Цілі числа (int)
age = 25
temperature = -10
big_number = 1_000_000  # можна використовувати _ для читабельності

In [43]:
print(big_number)

1000000


In [44]:
print(f"{big_number:_}")

1_000_000


Цілі числа в Python підтримують набір самих звичайних математичних операцій:
| операція | пояснення                     |
|----------|-------------------------------|
| x + y    | додавання                     |
| x - y    | віднімання                    |
| x * y    | множення                      |
| x / y    | ділення                       |
| x // y   | ціла частина від ділення      |
| x % y    | залишок від ділення           |
| -x       | зміна знаку числа             |
| x ** y   | підведення у ступінь          |

Операції мають пріоритет який можна змінювати використовуючи круглі дужки, усе так само як у математиці.

In [45]:
a = 4 + 3 * 2
b = (4 + 3) * 2
print(f'{a = }')
print(f'{b = }')

a = 10
b = 14


Зауважте, що результатом операції `/`, а також підведення у від'ємний або дійсний ступінь буде дійсне число.

In [46]:
print(20 / 3)

6.666666666666667


In [47]:
print(20 // 3)

6


In [48]:
print(20 % 3)

2


In [49]:
print(3 ** 4)

81


In [50]:
print(81 ** (1/4))

3.0


Також слід зазначити, що цілі числа в Python, на відміну від багатьох інших мов програмування, підтримують довгу арифметику.
В Python довга (або довільна) арифметика реалізована вбудовано для цілих чисел (тип `int`). Вам не потрібно писати спеціальний код для обробки дуже великих чисел, оскільки тип `int` автоматично підтримує числа будь-якого розміру. Просто використовуйте стандартні оператори, а Python автоматично керує розміром цілих чисел, дозволяючи їм зростати необмежено, на відміну від мов з фіксованою розмірністю цілих чисел. 

In [51]:
print(3 ** 150)

369988485035126972924700782451696644186473100389722973815184405301748249


Вище ми бачимо приклад того, як Python працює з **дуже великими цілими числами** — те, що називається **"arbitrary-precision integers"**.

---

### 🔍 Що відбувається в коді:

```python
print(8 ** 119)
```

- Підносимо число `8` до степеня `119`.
- Це створює **гігантське число**:  
  `369806799885152159729427087251456644481673063392722957731054430429731194129`

---

### 📐 Чому це можливо в Python?

На відміну від багатьох мов програмування (наприклад, C або Java), де цілі числа мають обмежений розмір (наприклад, `int32` або `int64`), Python автоматично переходить на **довільну точність**:

- Немає переповнення (`overflow`) — Python просто виділяє більше пам’яті.
- Тип `int` в Python — це **динамічна структура**, яка масштабується під потреби.

---

### 📊 Порівняння з іншими мовами:

| Мова       | Максимальний `int` (типовий) | Поведінка при переповненні |
|------------|------------------------------|-----------------------------|
| C          | 2,147,483,647 (`int32`)      | Переповнення, неправильне значення |
| Java       | 9,223,372,036,854,775,807 (`long`) | Переповнення |
| Python     | Без обмежень (`int`)         | Автоматичне розширення     |

---

### 🧠 Навіщо це потрібно?

- Криптографія (робота з великими простими числами)
- Наукові обчислення
- Теорія чисел
- Фінансові розрахунки з високою точністю

---

# 6.2 Дійсні числа (float)
У дійсних числах ціла частина від дробової відділяється знаком крапки.

In [52]:
# Дійсні числа
height = 175.5
temperature = -15.7
pi = 3.14159265359

Для запису дуже великих або дуже малих по модулю чисел використовують так званий запис «з плаваючою крапкою» (також науковий або інженерний запис). У цьому випадку число представляється у вигляді деякої десяткової дроби, яку називають мантисою, помноженої на цілочислений ступінь десяти (порядок). Наприклад, відстань від Землі до Сонця дорівнює `1.496×10¹¹`.

Числа с плавачою крапкою записуються так: спочатку пишеться `мантиса`, далі пишеться буква `e`, потім пишеться `порядок`. Пробіли всередині цього запису не ставляться.
```python
1.496e11
```

In [53]:
# Наукова нотація
scientific = 1.5e6   # 1500000.0
small = 2.5e-3       # 0.0025

In [54]:
print(f'{scientific = }')
print(f'{small = }')

scientific = 1500000.0
small = 0.0025


# 6.3 Точність обчислень

Дробові числа підтримують ті ж операції, що й цілі. Однак через особливості представлення чисел у пам'яті комп'ютера дробові числа неточні, це може спричиняти помилки при обчисленнях:

In [55]:
print(0.1 + 0.2)

0.30000000000000004


In [56]:
print(1/10 + 1/5)

0.30000000000000004


Це класична проблема **неточності з плаваючою комою** (floating-point precision), яка виникає через те, як комп’ютери представляють десяткові числа у **бінарному форматі**.

---

### 🔍 Що саме відбувається?

Коли ми виконуємо:

```python
print(0.1 + 0.2)
```

Очікуємо побачити `0.3`, але отримуємо `0.30000000000000004`. Причина — **деякі десяткові дроби не можуть бути точно представлені у двійковій системі**, так само як 1/3 не може бути точно записана у десятковій (0.333...).

У двійковій системі `0.1` і `0.2` — це нескінченні періодичні дроби, які комп’ютер **обрізає** до певної точності. Коли ці наближені значення додаються — виникає маленька похибка.

---

### 🧠 Чому це важливо?

- У фінансових розрахунках, фізичних симуляціях або тестах — навіть мікроскопічна похибка може призвести до неправильних результатів або провалу тестів.
- Це не баг Python — це **обмеження стандарту IEEE 754**, який використовують майже всі мови програмування.

---

Для високої точності використовують інші засоби (наприклад `Decimal` і `Fraction`)
Ось практичні приклади використання `Decimal` і `Fraction` — двох потужних інструментів з модуля `decimal` та `fractions`, які дозволяють уникнути проблем з точністю при роботі з числами.

### 🔢 `Decimal`: для точних десяткових обчислень

In [57]:
from decimal import Decimal, getcontext

getcontext().prec = 10  # встановлюємо точність до 10 знаків

a = Decimal('0.1')
b = Decimal('0.2')
c = a + b

print(c)  # 0.3

0.3


📌 **Переваги:**
- Точне представлення десяткових дробів
- Контроль над кількістю знаків після коми
- Ідеально для фінансових розрахунків

### 🧮 `Fraction`: для точних раціональних чисел

In [58]:
from fractions import Fraction

a = Fraction(1, 10)  # 0.1
print(a)

1/10


In [59]:
b = Fraction(1, 5)   # 0.2
print(b)

1/5


In [60]:
c = a + b
print(c)  # 3/10

3/10


In [61]:
print(type(c))

<class 'fractions.Fraction'>


In [62]:
print(float(c))  # 0.3

0.3


In [63]:
print(c + 1)

13/10


In [64]:
print(c + 0.1)

0.4


📌 **Переваги:**
- Зберігає точне співвідношення чисел (наприклад, 1/3)
- Уникає округлення до float
- Добре підходить для математичних задач, де важлива точність дробів

### 🔍 Порівняння:

| Засіб      | Тип точності       | Приклад використання         | Підходить для              |
|------------|--------------------|-------------------------------|----------------------------|
| `float`    | обмежена, неточна  | `0.1 + 0.2` → `0.30000000004` | загальні обчислення        |
| `Decimal`  | точна десяткова    | `Decimal('0.1') + Decimal('0.2')` → `0.3` | фінанси, звітність         |
| `Fraction` | точна раціональна | `Fraction(1, 10) + Fraction(1, 5)` → `3/10` | математика, теорія чисел   |

Також дійсні числа не підтримують довгу арифметику.

In [65]:
print(1/2 ** 50)

8.881784197001252e-16


In [66]:
print(2.0 + 2.0)

4.0


In [67]:
print(2 + 2.0)

4.0


# 6.4 Сomplex (комплексні числа)
Для математичних обчислень з комплексними числами:

In [68]:
# Комплексні числа
complex1 = 2 + 3j
print(complex1)

(2+3j)


In [69]:
print(type(complex1))

<class 'complex'>


In [70]:
complex2 = complex(4, -2)  # 4-2j
print(complex2)

(4-2j)


In [71]:
# Доступ до частин
print(complex2.real)   # 2.0 - дійсна частина

4.0


In [72]:
print(complex2.imag)   # 3.0 - уявна частина

-2.0


In [73]:
import cmath

complex3 = cmath.sqrt(-4)
print(complex3)

2j


# 7. Логічний тип (bool)

Може приймати лише два значення:

```python
# Булеві значення
is_student = True
is_finished = False

# Перетворення в bool
print(bool(1))      # True - будь-яке число крім 0
print(bool(0))      # False
print(bool(""))     # False - порожній рядок
print(bool("Hi"))   # True - непорожній рядок

In [74]:
# Булеві значення
is_student = True
is_finished = False

In [75]:
# Перетворення в bool
print(bool(1))      # True - будь-яке число крім 0

True


In [76]:
print(bool(0))      # False

False


In [77]:
print(bool(""))     # False - порожній рядок

False


In [78]:
print(bool("Hi"))   # True - непорожній рядок

True


In [79]:
print(2<3)

True


In [80]:
print(4>5)

False


In [81]:
print(bool(None))

False


# 8. `None` в Python: сутність "нічого"

В світі програмування на Python `None` — це спеціальний та унікальний об'єкт, який слугує для позначення відсутності значення. Це спосіб сказати, що змінна, параметр або результат функції не мають конкретного значення. `None` є єдиним екземпляром свого власного типу даних, `NoneType`.

Уявіть собі коробку. Вона може містити щось — число, рядок, список. А може бути порожньою. Ось ця "порожнеча" і є `None` в Python. Змінна існує, але вона не вказує на жоден конкретний об'єкт.

### Основні характеристики `None`:

  * **Єдиний екземпляр:** Всі посилання на `None` вказують на один і той самий об'єкт в пам'яті. Це робить перевірку на `None` дуже ефективною.
  * **Тип `NoneType`:** `None` є єдиним можливим значенням типу `NoneType`.
  * **"Хибне" значення (False):** В логічному контексті `None` поводиться як `False`. Це означає, що воно не пройде перевірку в умовному операторі `if`.

### Як правильно перевіряти на `None`?

Найкращий спосіб перевірити, чи є змінна `None` — це використання оператора `is`.

In [82]:
variable = None
print(variable is None)

True


In [83]:
print(type(variable))

<class 'NoneType'>


**Чому `is`, а не `==`?**

Хоча `змінна == None` в більшості випадків спрацює так само, як `is`, використання `is` є більш "пітонічним" (Pythonic) і надійним. Оператор `is` перевіряє, чи дві змінні вказують на один і той самий об'єкт в пам'яті. Оскільки `None` є унікальним об'єктом, ця перевірка є швидкою та гарантує, що ви порівнюєте саме з ним. Теоретично, об'єкт може бути створений таким чином, щоб `== None` повертало `True`, навіть якщо цей об'єкт не є справжнім `None`.

### `None`, `False` та `0`: в чому різниця?

Хоча всі три значення вважаються "хибними" (falsy) в логічному контексті (тобто `bool(None)`, `bool(False)` і `bool(0)` повернуть `False`), вони не є однаковими:

| Значення | Тип          | Опис                                  |
| :------- | :----------- | :------------------------------------ |
| `None`   | `NoneType`   | Представляє відсутність значення.     |
| `False`  | `bool`       | Представляє логічне значення "хиба".  |
| `0`      | `int` (або `float`) | Представляє числове значення "нуль". |

In [84]:
print(None == False)  # Виведе: False

False


In [85]:
print(None == 0)      # Виведе: False

False


In [86]:
print(False == 0)     # Виведе: True (з історичних причин)

True


Важливо розрізняти їх. Наприклад, функція може повернути `0`, що є валідним числовим результатом, або `None`, що може сигналізувати про помилку або відсутність даних. Використання `if` не дозволить розрізнити ці два випадки. Тому завжди краще явно перевіряти на `None` за допомогою `is`.

# 9. Символьні рядки `str` (*початкова інформація*)
Символьний рядок (або просто рядок, `string`) — це упорядкована послідовність символів яку використовують для зберігання та представлення текстової інформації.

Щоб створити рядок символи треба заключити у лапки або апострофи.

In [87]:
s = 'To be or not to be'
print(s)

To be or not to be


In [88]:
print(type(s))

<class 'str'>


In [89]:
s = "Основи програмування на Python"
print(s)

Основи програмування на Python


Якщо рядок містить, наприклад, лапки, тоді його треба заключити в апострофи і навпаки:

In [90]:
s = 'Вона сказала: "Нехай щастить"'
print(s)

Вона сказала: "Нехай щастить"


In [91]:
phrase = "It's cool!"
print(phrase)

It's cool!


In [92]:
complex_phrase = 'Це фраза "It\'s cool!"'
print(complex_phrase)

Це фраза "It's cool!"


У мові програмування **Python** символ зворотної скісної риски (`\`) виконує функцію **escape character** — спеціального маркера, який змінює інтерпретацію наступного за ним символа.  

У наведеному прикладі:
```python
complex_phrase = "Це фраза \"It's cool!\""
```
зворотна скісна риска перед подвійними лапками (`\"`) сигналізує інтерпретатору Python, що ці лапки **не є завершенням рядка**, а повинні бути включені до його вмісту як звичайний символ.

---

1. Проблема, яку вирішує екранування
- Рядки в Python можуть бути огорнуті **подвійними лапками** (`"..."`) або **одинарними лапками** (`'...'`).
- Якщо всередині рядка, який починається і закінчується подвійними лапками, з’являються ще одні подвійні лапки **без екранування**, інтерпретатор сприйме їх як кінець рядка, що призведе до **синтаксичної помилки**.

**Приклад помилки:**
```python
complex_phrase = "Це фраза "It's cool!""
# SyntaxError: invalid syntax
```

---

2. Механізм роботи `\`
- Символ `\` змінює значення наступного символа, дозволяючи вставити в рядок:
  - лапки (`\"` або `\'`)
  - спеціальні керівні символи (`\n` — новий рядок, `\t` — табуляція тощо)
  - символи, які інакше мали б інше синтаксичне значення.

У нашому випадку:
```python
\"  →  подвійна лапка, яка не завершує рядок, а стає його частиною
```

---

#### Альтернативи екрануванню
Екранування — не єдиний спосіб включити лапки в рядок:
1. **Використати інший тип лапок для огортання рядка**:
   ```python
   complex_phrase = 'Це фраза "It\'s cool!"'
   ```
2. **Використати потрійні лапки**:
   ```python
   complex_phrase = """Це фраза "It's cool!" """
   ```
---

### Висновок
Використання зворотної скісної риски у цьому прикладі є **необхідним синтаксичним прийомом** для коректного представлення символів, які мають спеціальне значення в синтаксисі Python. У даному випадку `\"` дозволяє вбудувати подвійні лапки у рядок, не порушуючи його структури. Це приклад **екранування символів** — фундаментального механізму в мовах програмування, що забезпечує точне відтворення текстових даних, які містять службові символи.

#### Рядки у потрійних лапках чи апострофах
Якщо треба визначити блок тексту який складається з декількох рядків, тоді такі послідовності символів заключають у потрйні лапки чи апострофи. Причому ми маємо змогу вільно користуватись звичайними лапками чи апострофами усередині таких блоків:

In [93]:
multiline = """Перший рядок
Другий рядок
Третій рядок"""
print(multiline)

Перший рядок
Другий рядок
Третій рядок


In [94]:
line3 = "Перший рядок\nДругий рядок\nТретій рядок"
print(line3)

Перший рядок
Другий рядок
Третій рядок


Базові операції над рядками: конкатенація (`+`), повторення (`*`), довжина (`len()`), екрановані послідовності (`\n`, `\t`, `\"`, тощо).

In [95]:
string1 = 'Hello'
string2 = 'World'
print(string1, string2)

Hello World


In [96]:
print(string1 + string2)

HelloWorld


In [97]:
print(string1 + ', ' + string2 + '!')

Hello, World!


In [98]:
print(string1 * 3)

HelloHelloHello


In [99]:
print(len(string1))

5


In [100]:
print('ll' in string1)

True


In [101]:
print('ll' in string2)

False


# 10. Операції в Python

# 10.1 Арифметичні операції

Python підтримує всі стандартні математичні операції:

| Оператор | Назва | Приклад | Результат | Примітка |
|----------|-------|---------|-----------|----------|
| `+` | Додавання | `10 + 5` | `15` | |
| `-` | Віднімання | `10 - 5` | `5` | |
| `*` | Множення | `10 * 5` | `50` | |
| `/` | Ділення | `10 / 4` | `2.5` | Завжди float |
| `//` | Цілочисельне ділення | `10 // 4` | `2` | Відкидає дробову частину |
| `%` | Остача від ділення | `10 % 3` | `1` | Модуль |
| `**` | Піднесення до степеня | `2 ** 3` | `8` | |

In [102]:
# Приклади арифметичних операцій
a = 10
b = 3

print(f"{a} + {b} = {a + b}")    # 10 + 3 = 13
print(f"{a} - {b} = {a - b}")    # 10 - 3 = 7
print(f"{a} * {b} = {a * b}")    # 10 * 3 = 30
print(f"{a} / {b} = {a / b}")    # 10 / 3 = 3.333...
print(f"{a} // {b} = {a // b}")  # 10 // 3 = 3
print(f"{a} % {b} = {a % b}")    # 10 % 3 = 1
print(f"{a} ** {b} = {a ** b}")  # 10 ** 3 = 1000

10 + 3 = 13
10 - 3 = 7
10 * 3 = 30
10 / 3 = 3.3333333333333335
10 // 3 = 3
10 % 3 = 1
10 ** 3 = 1000


In [103]:
# Пріоритет операцій (як у математиці)
result = 2 + 3 * 4  # Спочатку множення, потім додавання
print(result)  # 14

14


In [104]:
# Зміна пріоритету дужками
result = (2 + 3) * 4  # Спочатку дужки
print(result)  # 20

20


# 10.2 Складені оператори присвоєння

In [105]:
# Скорочені форми операторів
x = 10
x += 5   # x = x + 5, результат: 15
print(f'{x = }')

x = 15


In [106]:
x -= 3   # x = x - 3, результат: 12
print(f'{x = }')

x = 12


In [107]:
x *= 2   # x = x * 2, результат: 24
print(f'{x = }')

x = 24


In [108]:
x /= 4   # x = x / 4, результат: 6.0
print(f'{x = }')

x = 6.0


In [109]:
x //= 2  # x = x // 2, результат: 3.0
print(f'{x = }')

x = 3.0


In [110]:
x %= 2   # x = x % 2, результат: 1.0
print(f'{x = }')

x = 1.0


In [111]:
x **= 3  # x = x ** 3, результат: 1.0
print(f"Фінальне значення x: {x}")

Фінальне значення x: 1.0


# 10.3 Оператор "`walrus`" (`:=`)
або "моржовий оператор", у Python — це оператор присвоєння виразу, який дозволяє одночасно присвоювати значення змінній та використовувати це значення в тому ж виразі, на відміну від звичайного присвоєння (`=`), яке є окремою інструкцією. Він з'явився у Python 3.8 і допомагає зробити код більш лаконічним та читабельним, особливо в таких контекстах, як спискові включення, цикли while та умовні оператори. 
Як це працює 
1. Присвоєння значення:
Вираз зліва від `:=` (наприклад, `x`) обчислюється.
2. Присвоєння змінній:
Результат цього виразу присвоюється змінній `x`.
3. Повернення значення:
Сама змінна `x` повертає присвоєне їй значення.

In [112]:
print(y:=100)

100


In [113]:
y += 10
print(y)

110


#### Переваги оператора `walrus`

**Лаконічність**:

    Дозволяє скоротити код, уникаючи повторного обчислення виразів. 
    
**Зменшення повторень**:

    Зручно, коли значення використовується кілька разів у межах одного виразу або логічного блоку. 
    
**Читабельність**:

    Хоча спочатку може бути незвичним, при правильному використанні оператор робить код більш зрозумілим та структурованим.
    
Важливо пам'ятати: Оператор `:=` повертає присвоєне значення, що відрізняє його від звичайного оператора `=` і робить його корисним в контекстах, де потрібна як присвоєння, так і використання значення. 

# 10.4 Операції порівняння

Операції порівняння повертають логічне значення `True` або `False`:

| Оператор | Значення | Приклад | Результат |
|----------|----------|---------|-----------|
| `==` | Дорівнює | `5 == 5` | `True` |
| `!=` | Не дорівнює | `5 != 3` | `True` |
| `>` | Більше | `10 > 5` | `True` |
| `<` | Менше | `10 < 5` | `False` |
| `>=` | Більше або дорівнює | `5 >= 5` | `True` |
| `<=` | Менше або дорівнює | `10 <= 5` | `False` |

In [114]:
# Приклади операцій порівняння
x = 10
y = 20
z = 10

print(f"{x} == {y}: {x == y}")  # 10 == 20: False
print(f"{x} != {y}: {x != y}")  # 10 != 20: True
print(f"{x} < {y}: {x < y}")    # 10 < 20: True
print(f"{x} > {y}: {x > y}")    # 10 > 20: False
print(f"{x} <= {z}: {x <= z}")  # 10 <= 10: True
print(f"{x} >= {z}: {x >= z}")  # 10 >= 10: True

10 == 20: False
10 != 20: True
10 < 20: True
10 > 20: False
10 <= 10: True
10 >= 10: True


In [115]:
# Ланцюгове порівняння (особливість Python)
age = 25
is_adult = 18 <= age <= 65
print(f"Працездатний вік: {is_adult}")  # True

Працездатний вік: True


# 10.5 Логічні операції

Логічні оператори використовуються для комбінування логічних виразів:

| Оператор | Значення | Опис |
|----------|----------|------|
| `and` | Логічне І | `True`, якщо обидва операнди `True` |
| `or` | Логічне АБО | `True`, якщо хоча б один операнд `True` |
| `not` | Логічне НЕ | Інвертує логічне значення |


In [116]:
# Приклади логічних операцій
a = True
b = False

print(f"{a} and {b} = {a and b}")  # True and False = False
print(f"{a} or {b} = {a or b}")    # True or False = True
print(f"not {a} = {not a}")        # not True = False

True and False = False
True or False = True
not True = False


In [117]:
# Практичний приклад
age = 19
has_license = True
can_drive = (age >= 18) and has_license
print(f"Може керувати автомобілем: {can_drive}")  # True

Може керувати автомобілем: True


In [118]:
# Складні логічні вирази
temperature = 22
is_comfortable = (temperature >= 18) and (temperature <= 25)
print(f"Комфортна температура: {is_comfortable}")  # True

Комфортна температура: True


### Таблиці істинності
Демонстрація роботи логічних операторів

In [119]:
print("Таблиця істинності для AND:\n")
print(f"True and True = {True and True}")    # True
print(f"True and False = {True and False}")  # False  
print(f"False and True = {False and True}")  # False
print(f"False and False = {False and False}") # False

Таблиця істинності для AND:

True and True = True
True and False = False
False and True = False
False and False = False


In [120]:
print("Таблиця істинності для OR:\n")
print(f"True or True = {True or True}")      # True
print(f"True or False = {True or False}")    # True
print(f"False or True = {False or True}")    # True  
print(f"False or False = {False or False}")  # False

Таблиця істинності для OR:

True or True = True
True or False = True
False or True = True
False or False = False


In [121]:
print("Таблиця істинності для NOT:\n")
print(f"not True = {not True}")   # False
print(f"not False = {not False}") # True

Таблиця істинності для NOT:

not True = False
not False = True


# 10.6 Умовні операції (тернарний оператор)

In [122]:
# Тернарний оператор: значення_якщо_True if умова else значення_якщо_False
age = 17
status = "повнолітній" if age >= 18 else "неповнолітній"
print(f"Статус: {status}")  # Статус: неповнолітній

Статус: неповнолітній


In [123]:
# Еквівалентний if-else
if age >= 18:
    status = "повнолітній"
else:
    status = "неповнолітній"
print(f"Статус: {status}")

Статус: неповнолітній


# 11 Перетворення типів

# 11.1 Явне перетворення типів (Type Casting)

Python дозволяє явно перетворювати дані з одного типу в інший:

In [124]:
# Перетворення в ціле число int()
int_from_float = int(3.14)        # 3 (відкидає дробову частину)
int_from_string = int("42")       # 42
int_from_bool = int(True)         # 1
int_from_bool2 = int(False)       # 0

print(f"int(3.14) = {int_from_float}")
print(f"int('42') = {int_from_string}")
print(f"int(True) = {int_from_bool}")

int(3.14) = 3
int('42') = 42
int(True) = 1


In [125]:
# Перетворення в дійсне число float()
float_from_int = float(10)        # 10.0
float_from_string = float("3.14") # 3.14
float_from_bool = float(False)    # 0.0

print(f"float(10) = {float_from_int}")
print(f"float('3.14') = {float_from_string}")

float(10) = 10.0
float('3.14') = 3.14


In [126]:
# Перетворення в рядок str()
str_from_int = str(42)            # "42"
str_from_float = str(3.14)        # "3.14"
str_from_bool = str(True)         # "True"

print(f"str(42) = '{str_from_int}'")
print(f"str(3.14) = '{str_from_float}'")
print(f"str(True) = '{str_from_bool}'")

str(42) = '42'
str(3.14) = '3.14'
str(True) = 'True'


In [127]:
# Перетворення в логічний тип bool()
bool_from_int = bool(0)           # False (тільки 0 = False)
bool_from_int2 = bool(42)         # True (будь-яке інше число = True)
bool_from_string = bool("")       # False (порожній рядок = False)
bool_from_string2 = bool("text")  # True (непорожній рядок = True)

print(f"bool(0) = {bool_from_int}")
print(f"bool(42) = {bool_from_int2}")
print(f"bool('') = {bool_from_string}")
print(f"bool('text') = {bool_from_string2}")

bool(0) = False
bool(42) = True
bool('') = False
bool('text') = True


#### Значення, що вважаються False:

In [128]:
# В Python наступні значення вважаються False при перетворенні в bool:
falsy_values = [
    0,           # ноль
    0.0,         # нуль з плаваючою крапкою  
    "",          # порожній рядок
    [],          # порожній список
    (),          # порожній кортеж
    {},          # порожній словник
    set(),       # порожня множина
    None         # None
]

In [129]:
for value in falsy_values:
    print(f"bool({value}) = {bool(value)}")  # Всі дають False

bool(0) = False
bool(0.0) = False
bool() = False
bool([]) = False
bool(()) = False
bool({}) = False
bool(set()) = False
bool(None) = False


# 11.2 Неявне перетворення типів

Python автоматично перетворює менші типи в більші для уникнення втрати даних:

In [130]:
# Неявне перетворення при арифметичних операціях
integer_num = 10      # int
float_num = 3.5       # float
result = integer_num + float_num  # int + float = float
print(f"Тип результату: {type(result)}, значення: {result}")  # float, 13.5

Тип результату: <class 'float'>, значення: 13.5


In [131]:
# Більше прикладів неявного перетворення
bool_val = True       # bool
int_val = 5          # int
result2 = bool_val + int_val  # bool + int = int (True = 1)
print(f"True + 5 = {result2}")  # 6

True + 5 = 6


In [132]:
# При множенні рядка на число
text = "Python"      # str
count = 3            # int  
repeated = text * count  # str * int = str
print(f"'{text}' * {count} = '{repeated}'")  # 'PythonPythonPython'

'Python' * 3 = 'PythonPythonPython'


Неявне (автоматичне) приведення типів для чисел у Python — це один з найяскравіших прикладів `type coercion`. Python автоматично перетворює числові типи для забезпечення сумісності операцій.

# 11.3 Ієрархія числових типів

Python дотримується чіткої ієрархії при приведенні числових типів:


`bool` → `int` → `float` → `complex`


**Правило:** менш "складний" тип завжди приводиться до більш "складного".

In [133]:
# bool + int
result = True + 5
print(result)        # 6
print(type(result))  # <class 'int'>

6
<class 'int'>


In [134]:
result = False * 10
print(result)        # 0
print(type(result))  # <class 'int'>

0
<class 'int'>


In [135]:
# int + float
result = 3 + 2.5
print(result)        # 5.5
print(type(result))  # <class 'float'>

5.5
<class 'float'>


In [136]:
result = 10 / 3      # Ділення завжди повертає float
print(result)        # 3.3333333333333335
print(type(result))  # <class 'float'>

3.3333333333333335
<class 'float'>


In [137]:
# float + complex
result = 2.5 + (1 + 2j)
print(result)        # (3.5+2j)
print(type(result))  # <class 'complex'>

(3.5+2j)
<class 'complex'>


In [138]:
result = 3.14 * 2j
print(result)        # 6.28j
print(type(result))  # <class 'complex'>

6.28j
<class 'complex'>


#### Комбіновані операції

In [139]:
# bool + int + float + complex
result = True + 2 + 3.5 + (1 + 1j)
print(result)        # (7.5+1j)
print(type(result))  # <class 'complex'>

(7.5+1j)
<class 'complex'>


#### Операції порівняння

При порівнянні Python також виконує приведення типів:

In [140]:
print(1 == 1.0)      # True

True


In [141]:
print(True == 1)     # True

True


In [142]:
print(False == 0.0)  # True

True


In [143]:
print(2 == (2+0j))   # True

True


In [144]:
# Порівняння величини
print(1 < 2.5)       # True

True


In [145]:
print(True < 3.14)   # True

True


#### Детальний розбір прикладів

##### Приклад 1: Послідовне приведення

In [146]:
a = True     # bool
b = 5        # int  
c = 2.3      # float
d = 1 + 1j   # complex

step1 = a + b        # True + 5 → 1 + 5 = 6 (int)
step2 = step1 + c    # 6 + 2.3 → 6.0 + 2.3 = 8.3 (float)  
step3 = step2 + d    # 8.3 + (1+1j) → (8.3+0j) + (1+1j) = (9.3+1j) (complex)

print(f"Фінальний результат: {step3}")  # (9.3+1j)
print(f"Тип: {type(step3)}")             # <class 'complex'>

Фінальний результат: (9.3+1j)
Тип: <class 'complex'>


#### Приклад 2: Математичні функції

In [147]:
import math

# Функції можуть приймати різні числові типи
print(math.sqrt(4))      # 2.0 (int → float)

2.0


In [148]:
print(math.sqrt(4.0))    # 2.0 (вже float)

2.0


In [149]:
print(math.sqrt(True))   # 1.0 (bool → float)

1.0


In [150]:
# Степінь може повертати різні типи
print(2 ** 3)            # 8 (int)

8


In [151]:
print(2 ** 3.0)          # 8.0 (float)

8.0


In [152]:
print(2.0 ** 3)          # 8.0 (float)

8.0


#### Спеціальні випадки

##### Ділення

In [153]:
# Звичайне ділення завжди повертає float
print(6 / 2)             # 3.0 (не 3!)
print(type(6 / 2))       # <class 'float'>

3.0
<class 'float'>


In [154]:
# Цілочисельне ділення зберігає тип
print(6 // 2)            # 3
print(type(6 // 2))      # <class 'int'>

3
<class 'int'>


In [155]:
# Але може приводити до float
print(7.0 // 2)          # 3.0
print(type(7.0 // 2))    # <class 'float'>

3.0
<class 'float'>


Автоматичне приведення числових типів у Python робить мову гнучкою та зручною, але розуміння цих механізмів допомагає писати більш передбачуваний код.

# 12 Консольне введення інформації

# 12.1 Функція input()

Функція `input()` використовується для отримання даних від користувача через консоль:

In [156]:
# Базове введення (завжди повертає рядок!)
name = input("Введіть ваше ім'я: ")
print(f"Привіт, {name}!")

Введіть ваше ім'я:  Serhii


Привіт, Serhii!


In [157]:
# ВАЖЛИВО: input() завжди повертає рядок
age_str = input("Введіть ваш вік: ")
print(type(age_str))  # <class 'str'> - це рядок!

Введіть ваш вік:  40


<class 'str'>


### Введення різних типів даних
Оскільки `input()` повертає рядок, потрібне перетворення типів:

In [158]:
# Введення цілого числа
age = int(input("Введіть ваш вік: "))
year_of_birth = 2025 - age
print(f'Ви наробилися в {year_of_birth} році')

Введіть ваш вік:  40


Ви наробилися в 1985 році


In [159]:
# Введення логічного значення
answer = input("Ви студент? (так/ні): ")
is_student = answer.lower() == "так"  # True або False
print(f'Користувач {name} {("" if is_student else "не ")}студент')

Ви студент? (так/ні):  ні


Користувач Serhii не студент


In [160]:
# Введення кількох значень в одному рядку
data = input("Введіть три числа через пробіл: ")  # "10 20 30"
numbers = data.split()  # ['10', '20', '30']
a, b, c = numbers
print(f'{a = }; {b = }; {c = };')
summa = int(a) + int(b) + int(c)
print(f'{summa = }')

Введіть три числа через пробіл:  10 20 30


a = '10'; b = '20'; c = '30';
summa = 60


In [161]:
print(f'summa = {sum(map(int, (data:=input("Введіть три числа через пробіл: ")).split()))}')

Введіть три числа через пробіл:  10 20 30


summa = 60


### Покрокова розбивка:

`input("Введіть три числа через пробіл: ")` - виводить повідомлення і чекає на введення користувача

`(data := ...)` - це "walrus operator" (оператор морж) - присвоює результат input() змінній data і одночасно повертає це значення для подальшого використання

`.split()` - розділяє введений рядок за пробілами, повертаючи список рядків

`map(int, ...)` - застосовує функцію int() до кожного елемента списку, перетворюючи рядки в числа

`sum(...)` - обчислює суму всіх чисел

`f'summa = {...}'` - f-string для форматування виводу, вставляє результат суми в рядок

`print(...)` - виводить фінальний результат

# 12.2 Функція print()

Базове виведення даних:

In [162]:
# Простий вивід
print("Привіт, Світ!")

Привіт, Світ!


In [163]:
# Вивід кількох значень
name = "Олексій"
age = 22
print("Ім'я:", name, "\nВік:", age)

Ім'я: Олексій 
Вік: 22


In [164]:
# Параметри функції print()
print("Рядок 1", end=" ")  # без переходу на новий рядок
print("Рядок 2")           # Виведе: Рядок 1 Рядок 2

Рядок 1 Рядок 2


In [165]:
print("A", "B", "C", sep="-")  # Виведе: A-B-C

A-B-C


# 12.3 Форматоване виведення з f-рядками

# 12.3.1 Основи f-рядків

**F-рядки** (f-strings) — найсучасніший і зручний спосіб форматування рядків у Python:

In [166]:
# Базове використання f-рядків  
name = "Марія"
age = 25
city = "Львів"

# Просте підставлення значень
greeting = f"Привіт! Мене звати {name}, мені {age} років"
print(greeting)

Привіт! Мене звати Марія, мені 25 років


In [167]:
# Можна використовувати вирази всередині {}
print(f"Наступного року мені буде {age + 1} років")
print(f"Я живу в місті {city.upper()}")

Наступного року мені буде 26 років
Я живу в місті ЛЬВІВ


In [168]:
info = f"""Особиста інформація:
Ім'я: {name}
Вік: {age} років
Місто: {city}"""
print(info)

Особиста інформація:
Ім'я: Марія
Вік: 25 років
Місто: Львів


# 12.3.2 Форматування чисел

In [169]:
# Форматування дійсних чисел
pi = 3.14159265359
price = 1234.567
percentage = 0.856

In [170]:
# Кількість знаків після коми
print(f"π з 2 знаками: {pi:.2f}")        # 3.14
print(f"π з 4 знаками: {pi:.4f}")        # 3.1416
print(f"π без дробової частини: {pi:.0f}") # 3

π з 2 знаками: 3.14
π з 4 знаками: 3.1416
π без дробової частини: 3


In [171]:
# Форматування грошових сум
print(f"Ціна товару: {price:.2f} грн")    # 1234.57 грн

Ціна товару: 1234.57 грн


In [172]:
# Відсотки
print(f"Успішність: {percentage:.1%}")    # 85.6%
print(f"Успішність: {percentage:.0%}")    # 86%

Успішність: 85.6%
Успішність: 86%


In [173]:
# Наукова нотація
big_number = 1234567890
print(f"Велике число: {big_number:.2e}")  # 1.23e+09

Велике число: 1.23e+09


In [174]:
# Розділювачі тисяч
large_sum = 1234567
print(f"З розділювачами: {large_sum:,}")  # 1,234,567 (на англійських системах)

З розділювачами: 1,234,567


In [175]:
# Форматування грошових сум
salary = 12500.50
print(f"Зарплата: {salary:,.2f} грн")  # 12,500.50 грн

Зарплата: 12,500.50 грн


# 12.3.3 Вирівнювання та заповнення

In [176]:
# Вирівнювання тексту
product = "Python"
price = 199.99

In [177]:
# Ліворуч (<), по центру (^), праворуч (>)
print(f"Товар: |{product:<15}| Ціна: |{price:>8.2f}|")
print(f"Товар: |{product:^15}| Ціна: |{price:^8.2f}|")  
print(f"Товар: |{product:>15}| Ціна: |{price:<8.2f}|")

Товар: |Python         | Ціна: |  199.99|
Товар: |    Python     | Ціна: | 199.99 |
Товар: |         Python| Ціна: |199.99  |


In [178]:
# Заповнення нулями
order_number = 42
print(f"Номер замовлення: {order_number:05}")  # 00042

Номер замовлення: 00042


In [179]:
# Заповнення іншими символами  
title = "МЕНЮ"
print(f"{title:=^30}")      # ============МЕНЮ=============
print(f"{title:*<30}")      # МЕНЮ**************************
print(f"{title:-^30}")      # ------------МЕНЮ-------------

=============МЕНЮ=============
МЕНЮ**************************
-------------МЕНЮ-------------


# 13 Вбудовані функції для перевірки типів і діапазонів

# 13.1 Функції для перевірки типів

In [180]:
# Функція type() - повертає тип об'єкта
number = 42
text = "Python"
flag = True

print(f"type({number}) = {type(number)}")  # <class 'int'>
print(f"type('{text}') = {type(text)}")    # <class 'str'>  
print(f"type({flag}) = {type(flag)}")      # <class 'bool'>

type(42) = <class 'int'>
type('Python') = <class 'str'>
type(True) = <class 'bool'>


In [181]:
# isinstance() - більш гнучка перевірка типу
value = 42

# Перевірка одного типу
print(isinstance(value, int))      # True

True


In [182]:
print(isinstance(value, str))      # False

False


In [183]:
# Перевірка кількох типів одночасно
number = 3.14
print(isinstance(number, (int, float)))  # True - або int, або float

True


In [184]:
print(isinstance(value, (int, float)))  # True - або int, або float

True


In [185]:
value = 42
print(f'{value = } є цілим числом: {isinstance(value, int)}')

value = 42 є цілим числом: True


In [186]:
value = 3.14
print(f'{value = } є дійсним числом: {isinstance(value, float)}')

value = 3.14 є дійсним числом: True


In [187]:
value = "Hello"
print(f'{value = } є рядком: {isinstance(value, str)}')

value = 'Hello' є рядком: True


In [188]:
value = True
print(f'{value = } є логічним значенням: {isinstance(value, bool)}')

value = True є логічним значенням: True


In [189]:
value = 10
print(f'{value = } є числом (int або float): {isinstance(value, (int, float))}')

value = 10 є числом (int або float): True


# 13.2 Функції для роботи з числами

In [190]:
# min() та max() - мінімальне та максимальне значення
minimum = min(10, 5, 15)
print(minimum)

5


In [191]:
maximum = max(10, 5, 15)
print(maximum)

15


In [192]:
# abs() - абсолютне значення (модуль числа)
negative = -15
print(f"abs({negative}) = {abs(negative)}")   # 15

abs(-15) = 15


In [193]:
# sum() - сума елементів послідовності
summa = sum((10, 5, 15))
print(f"Сума всіх чисел: {summa}")    # 20.9

Сума всіх чисел: 30


In [194]:
# pow() - піднесення до степеня
print(pow(2, 3))      # 8 (еквівалентно 2**3)
print(2 ** 3)

8
8


In [195]:
# round() - округлення
pi = 3.14159
print(f"round({pi}) = {round(pi)}")           # 3
print(f"round({pi}, 2) = {round(pi, 2)}")     # 3.14
print(f"round({pi}, 4) = {round(pi, 4)}")     # 3.1416

round(3.14159) = 3
round(3.14159, 2) = 3.14
round(3.14159, 4) = 3.1416


In [196]:
# Особливість round() - округлення до найближчого парного
print(f"round(2.5) = {round(2.5)}")  # 2
print(f"round(3.5) = {round(3.5)}")  # 4

round(2.5) = 2
round(3.5) = 4


У Python вбудована `round()` не завжди округляє “в більшу” при `.5` — вона використовує так зване “банківське округлення” (`round half to even`).

🔹 Як це працює:
- Якщо дробова частина менше `0.5` → `округлення вниз`.
- Якщо більше `0.5` → `округлення вгору`.
- Якщо рівно `0.5` → `округлення до найближчого парного цілого`.


In [197]:
print(f"round(2.51) = {round(2.5)}")  # 2
print(f"round(3.49) = {round(3.5)}")  # 4

round(2.51) = 2
round(3.49) = 4


💡 Якщо потрібно завжди округляти в більшу сторону, навіть при .5, використовуйте наступне:

In [198]:
print(f"math.ceil(2.5) = {math.ceil(2.5)}")  # 2
print(f"math.ceil(3.5) = {math.ceil(3.5)}")  # 4

math.ceil(2.5) = 3
math.ceil(3.5) = 4


🔹 Типи округлення

- `round()` — банківське округлення (до найближчого парного при .5).
- `math.ceil()` — завжди вгору (до більшого цілого).
- `math.floor()` — завжди вниз (до меншого цілого).


In [199]:
# Додатні числа
print(math.floor(3.9))   # 3   (вниз до меншого цілого)
print(math.floor(3.1))   # 3
print(math.floor(3.0))   # 3   (ціле залишається без змін)

3
3
3


In [200]:
# Додатні числа
print(math.ceil(3.9))   # 4   (вгору до більшого цілого)
print(math.ceil(3.1))   # 4
print(math.ceil(3.0))   # 3   (ціле залишається без змін)

4
4
3


In [201]:
# Значення з .5
print(math.floor(2.5))   # 2
print(math.floor(7.5))   # 7

2
7


In [202]:
# Значення з .5
print(math.ceil(2.5))   # 3
print(math.ceil(7.5))   # 8

3
8


In [203]:
# Від’ємні числа
print(math.floor(-1.2))  # -2  (вниз — ще менше)
print(math.floor(-1.9))  # -2
print(math.floor(-3.0))  # -3  (ціле залишається без змін)
print(math.floor(-3.5))  # -4  (вниз — ще менше)


-2
-2
-3
-4


In [204]:
# Від’ємні числа
print(math.ceil(-1.2))  # -1  (вгору — тобто ближче до нуля)
print(math.ceil(-1.9))  # -1
print(math.ceil(-3.0))  # -3  (ціле залишається без змін)
print(math.ceil(-3.5))  # -3  (вгору — ближче до нуля)

-1
-1
-3
-3


🔹 Логіка:
- Для додатних значень “вгору” означає більше або рівне ціле число праворуч на числовій осі.
- Для від’ємних значень “вгору” означає рух у бік нуля (менше за модулем).

# Висновок до Теми 3

У межах теми сформовано системне розуміння фундаментів мови Python, що включає структуру програм, синтаксис, модель даних, операції, механізми введення/виведення та форматування результатів. Це забезпечує методологічну основу для подальшого вивчення структур даних, оброблення винятків, модульності та тестування.

1) Структура та синтаксис
- Розглянуто типову організацію коду: імпорти, константи, функції, точка входу `if __name__ == "__main__":` і роль документаційних рядків (docstring) та коментарів для відтворюваності й підтримуваності коду.
- Наголошено на відступах як маркерах блоків коду, регістрочутливості і практиках читабельності (перенесення в дужках замість \\).

2) Змінні, іменування, форматування
- Уточнено поняття змінної, ініціалізацію, множинне присвоєння, обмін значеннями, стилістичні норми іменування (`snake_case`).
- Показано сучасні підходи до форматування рядків: перевага f-рядків, а також `%` і `str.format()` для сумісності.

3) Об’єктна модель і типи
- Систематизовано три характеристики об’єкта: ідентичність, тип, значення; розмежовано `==` (значення) і `is` (ідентичність), пояснено інтернування малих цілих.
- Розглянуто динамічну типізацію і анотації типів (type hints) як інструмент підвищення якості коду; наведено базові типи (`int`, `float`, `str`, `bool`, `None`).

4) Числові типи і точність
- Деталізовано роботу з цілими (у т.ч. довільна точність), дійсними (особливості IEEE 754 і похибки 0.1+0.2), комплексними числами.
- Показано способи підвищення точності: `Decimal`, `Fraction`; розкрито арифметичні операції, пріоритети, округлення (`round` із банківською логікою, `math.ceil`, `math.floor`).

5) Рядки і базові операції
- Наведено створення рядків у різних лапках, екранування, багаторядкові літерали; основні операції: конкатенація, множення, довжина, пошук підрядка.

6) Логічні значення, None і логічна алгебра
- Узагальнено `bool`, семантику `None` та коректні перевірки через `is`; таблиці істинності для `and`, `or`, `not`; ланцюгові порівняння.

7) Операції перетворення типів (casting) і неявні перетворення
- Продемонстровано явні перетворення `int()`, `float()`, `str()`, `bool()` та набір значень, що трактуються як `False`.
- Пояснено неявне приведення у числовій ієрархії `bool → int → float → complex` і наслідки для арифметики та порівнянь.

8) Введення/виведення та форматування результатів
- Розглянуто `input()` (рядкову природу результату, потребу в перетворенні типів), читання кількох значень, використання оператора `:=` (walrus) у виразах.
- Поглиблено `print()` і параметри `sep`, `end`; f-рядки для чисел: фіксована точність, відсотки, наукова нотація, розділювачі тисяч, вирівнювання й заповнення.

9) Діагностика типів і базові числові функції
- Застосування `type()`, `isinstance()` для перевірки типів; `min`, `max`, `abs`, `sum`, `pow`, `round`, а також `math.ceil`, `math.floor` з прикладами на додатних і від’ємних значеннях.

Підсумок: Матеріал лекції формує цілісну когнітивну опору для грамотного використання Python у прикладних завданнях: від правильної організації коду й дисципліни синтаксису до свідомого керування типами, точністю обчислень і якісного введення/виведення. Набуті знання безпосередньо масштабуються на наступні теми — колекції, оброблення винятків, модульну організацію проєктів та тестування.
